Lesson3 (Container adapters)/FunctionsForTest.cpp

```diff
@@ -0,0 +1,66 @@
+#include <iostream>
+#include <stack>
+#include <queue>
+#include <memory>
+#include "FunctionsForTest.h"
+
+bool isExpressionValid (const std::string& expression)
+{
+    std::stack<char> MyStack;
+    char popStack;
+    bool result = true;
+    for(const auto& element : expression)
+    {
+        if ((element == '[')||(element == '(')||(element =='{'))
+        {
+            MyStack.push(element);
+        }
+        else
+        {
+            if(MyStack.empty())
+            {
+                result = false;
+                break;
+            }
+            popStack = MyStack.top();
+            MyStack.pop();
+            if (!(((element == ')')&&(popStack=='('))||
+                    ((element == '}')&&(popStack=='{'))||
+                    ((element == ']')&&(popStack=='['))))
+            {
+                result = false;
+                break;
+            }
+
+        }
+
+    }
+    if(!MyStack.empty())
+    {
+        result = false;
+    }
+    return result;
+}
+
+void linkSamelevel(std::shared_ptr<node> t)
+{
+    std::queue <std::shared_ptr<node>> q;
```

Comment:
**Not accepted. It's a copy-paste from https://www.geeksforgeeks.org/connect-nodes-level-level-order-traversal/.**

Reply:
So, here is another task for you:
Implement binary tree wrapper class.
You should implement add/pop (by key and by iterator)/clear methods.
It must support level order traversal with iterators (so, you must implement your own iterator).
With this class, write 15 unit tests for it (and cover the main public operations).

This task must be done till Thursday 2.00 PM (03 October).